src/movies.js

```diff
@@ -1,16 +1,140 @@
 // Iteration 1: All directors? - Get the array of all directors.
+function getAllDirectors(movieArr){
+    const listDirectors = movieArr.map(item =>  `${item.director}`)
+    return listDirectors
+  } 
+
 // _Bonus_: It seems some of the directors had directed multiple movies so they will pop up multiple times in the array of directors. How could you "clean" a bit this array and make it unified (without duplicates)?
+function getAllDirectors(movieArr){
+    const listDirectors = movieArr.map(item =>  `${item.director}`)
+    const listDirectorsfilter = listDirectors.filter((item, index) => {
+      return listDirectors.indexOf(item) === index
+    })
+    return listDirectorsfilter
+   }
+   
+   getAllDirectors(movies)  
 
 // Iteration 2: Steven Spielberg. The best? - How many drama movies did STEVEN SPIELBERG direct?
+function howManyMovies (arr) {
+    const steven = arr.filter (movie => movie.director === 'Steven Spielberg' && movie.genre.includes('Drama'))
+    return steven.length 
+  }
+  howManyMovies(movies)
+
 
 // Iteration 3: All rates average - Get the average of all rates with 2 decimals
+function ratesAverage (arr) {
+    if (arr.length === 0){
+      return 0
+    } 
+    
+    else {
+    const rateArr = arr.map(item => item.rate)
+    console.log(rateArr)
+
+    const onlyActualRatesArr = rateArr.filter(element => typeof element === 'number')
+      
+    const sumOfRates = onlyActualRatesArr.reduce((acc, c) =>  acc + c);
+    const averageRates = sumOfRates/rateArr.length;
+    return Number(averageRates.toFixed(2));
+  }
+}
+  
+  ratesAverage(movies)
+
 
 // Iteration 4: Drama movies - Get the average of Drama Movies
+function dramaMoviesRate(arr){
+  
+    const listGenre = arr.filter(movie => movie.genre.includes("Drama"))
+    //return listGenre  
+    if (listGenre.length == 0){
+        return 0;
+    }
+    
+    const rateArr = listGenre.map(item => item.rate)
+    const sumOfRates = rateArr.reduce((acc, c) =>  acc + c);
+    return Number((sumOfRates/rateArr.length).toFixed(2));
+  }
+  
+  dramaMoviesRate(movies)
 
 // Iteration 5: Ordering by year - Order by year, ascending (in growing order)
+function orderByYear (arr){
+    let byYear = []
+    const sortedYears = arr.sort((a, b) => {
+      let alpha = a.year - b.year 
+      if (alpha === 0){
+        return a.title.localeCompare(b.title)
+      }
+      return alpha
+     }) 
+     byYear = sortedYears
+    return byYear
+}
+    
+orderByYear(movies)
+
+
+// failed attempt to fix error (should return new array) in Jasmine
```

Comment:
When using sort() your are mutating the original array so you should create a safe copy of the array before sorting, for example: 
`function orderByYear(arr) {`
  `let byYear = [...arr];`
`  byYear.sort((a, b) => {`
   ` let alpha = a.year - b.year `
    `if (alpha === 0){`
   `   return a.title.localeCompare(b.title)`
  `  }`
`    return alpha`
  ` }) `
`  return byYear;`
`}`